blocks/conditional-panel-color/index.js

```diff
@@ -0,0 +1,24 @@
+/**
+ * External dependencies
+ */
+import { isEmpty } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { ifCondition, PanelColor, withContext } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+
+export default compose(
```

Comment:
Was trying to think of a way we could avoid duplicating the logic here between `ColorPalette` and `ConditionalPanelColor`. Unfortunate there's no easy way to avoid rendering a wrapper if none of its children render. I wonder if we could do a render prop which self-contains the null rendering:

```jsx
<EditorColors>
	{ ( { colors, disableCustomColors } ) => (
		<PanelColor title={ __( 'Text Color' ) } colorValue={ textColor }>
			<ColorPalette
				value={ textColor }
				onChange={ ( colorValue ) => setAttributes( { textColor: colorValue } ) }
				colors={ colors }
				disableCustomColors={ disableCustomColors }
			/>
		</PanelColor>
	) }
</EditorColors>
```

```js
const EditorColors = compose(
	withContext( 'editor' )(
		( settings, props ) => ( {
			colors: props.colors || settings.colors,
			disableCustomColors: props.disableCustomColors !== undefined ?
				props.disableCustomColors :
				settings.disableCustomColors,
		} )
	),
	ifCondition(
		( { colors, disableCustomColors } ) => ! ( isEmpty( colors ) && disableCustomColors )
	)
)( ( { children, ...props } ) => children( props ) );
```

This way, we don't need `ColorPalette` to retrieve color settings from context, because they'd always be passed in.

Thoughts?

Reply:
In either case, if we're creating a new component, it should be added to `blocks/index.js` to export on the global.

Reply:
Hi @aduth thank you for providing an alternative implementation, I really like the EditorColors component idea. I applied it to this PR.

Reply:
While I see the value of avoiding duplication, I think from a plugin author's perspective, duplication is fine or at least we could extract the logic into a `EditorColors` component but I think it should be baked in into these two components instead of using it at the block level. The block author could just use the `ColorPalette` or `ColorPanel` directly in that case.

Reply:
Hi @youknowriad, I understand your points. PanelColor is a generic component in @wordpress /components and we should not add the context information or conditional render directly to it. Maybe we should keep the context in the ColorPalette but change it to use a component that retrieves colors from the context. 
Then we create a new component HasColorsColorsToChoose or something similar that conditional render the children if colors are available or custom colors are enabled. And we use it to wrap color logic.
The other alternative is similar to what I had at the start a ConditionalPanelColor component that just renders PanelColor , but now we make use of a component that retrieves the colors from the state so no code duplication exists when comparing to the ColoPalette.

Reply:
>The other alternative is similar to what I had at the start a ConditionalPanelColor component that just renders PanelColor , but now we make use of a component that retrieves the colors from the state so no code duplication exists when comparing to the ColoPalette.

I was thinking more in that vein. The same we we have `wp.blocks.ColorPalette` we could have `we.blocks.ColorPanel` with context awareness. We can push the logic further by extracting the generic Color Palette to the components module so we'd also have `wp.components.ColorPalette` and `wp.components.ColorPanel` without any context awareness. Thoughs?

Reply:
I think it makes sense but maybe in the @blocks version, we should remove support for passing colors as props when that is a requirement the @components version can be used.
As if the behavior of the PanelColor in blocks is to not render if no colors are available if the plugin creator wants custom colors using the @blocks he would need to pass the colors also PanelColor which would only use them to check the conditional logic and does not need to know the colors for anything else.

Reply:
makes sense to me 👍 just thinking how we'd ensure backwards compatibility in that case. Maybe support it for a release or two with a deprecated message?

Reply:
Exactly, for the PanelColor @blocks  no deprecation is necessary as it is something new the existing one is at @components and will stay untouched. For the ColorPalette we would support but with a deprecation message when the prop is used saying to change to the @components version.